### ISSUE TITLE
tflite nightly/2.9.1 on iOS (CocoaPods) linker/build issue

### ISSUE THREAD

**Issue by sschaetz** [tensorflow/tensorflow#56255]
### Issue Type

Build/Install

### Source

binary

### Tensorflow Version

2.9.1 and 0.0.1-nightly.20220524

### Custom Code

No

### OS Platform and Distribution

MacOS 12.1, XCode 13.4, CocoaPod 1.11.3

### Mobile device

iOS 15

### Python version

n/a

### Bazel version

n/a

### GCC/Compiler version

n/a

### CUDA/cuDNN version

n/a

### GPU model and memory

n/a

### Current Behaviour?

Compiling a simple iOS project with this podfile:

```shell
platform :ios, '15.5'

target 'my-app' do
  use_frameworks!

  # Pods for my-app
  pod 'TensorFlowLiteSwift', '~> 0.0.1-nightly'
  pod 'TensorFlowLiteSelectTfOps', '~> 0.0.1-nightly'
end
```

and adding `-force_load "$(PROJECT_DIR)Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.xcframework/ios-arm64/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps"`

results in linker error (see below).

### Standalone code to reproduce the issue
n/a

### Relevant log output

```shell
Ld /Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos/my-app.app/my-app normal (in target 'my-app' from project 'my-app')
    cd /Users/me/dev/cmc/mobile-prototypes/my-app
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target arm64-apple-ios15.5 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk -L/Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphoneos -L/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk/usr/lib/swift -F/Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos -F/Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos/TensorFlowLiteSwift -F/Users/me/dev/cmc/mobile-prototypes/my-app/Pods/TensorFlowLiteC/Frameworks -F/Users/me/dev/cmc/mobile-prototypes/my-app/Pods/TensorFlowLiteSelectTfOps/Frameworks -F/Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos/XCFrameworkIntermediates/TensorFlowLiteC/Core -F/Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos/XCFrameworkIntermediates/TensorFlowLiteSelectTfOps -filelist /Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Intermediates.noindex/my-app.build/Debug-iphoneos/my-app.build/Objects-normal/arm64/my-app.LinkFileList -Xlinker -rpath -Xlinker /usr/lib/swift -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -rpath -Xlinker @executable_path/Frameworks -dead_strip -Xlinker -object_path_lto -Xlinker /Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Intermediates.noindex/my-app.build/Debug-iphoneos/my-app.build/Objects-normal/arm64/my-app_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -fembed-bitcode-marker -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphoneos -L/usr/lib/swift -Xlinker -add_ast_path -Xlinker /Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Intermediates.noindex/my-app.build/Debug-iphoneos/my-app.build/Objects-normal/arm64/beaker_proto2.swiftmodule -force_load /Users/me/dev/cmc/mobile-prototypes/my-app/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.xcframework/ios-arm64/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps -framework Pods_beaker_proto2 -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Intermediates.noindex/my-app.build/Debug-iphoneos/my-app.build/Objects-normal/arm64/my-app_dependency_info.dat -o /Users/me/Library/Developer/Xcode/DerivedData/my-app-hfikcxuyugearxghudrfwmrazkdp/Build/Products/Debug-iphoneos/my-app.app/my-app

ld: warning: Could not find or use auto-linked framework 'TensorFlowLiteC'
Undefined symbols for architecture arm64:
  "_TfLiteModelCreateFromFile", referenced from:
      TensorFlowLite.Model.init(filePath: Swift.String) -> TensorFlowLite.Model? in TensorFlowLite(Model.o)
  "_TfLiteSignatureRunnerGetOutputTensor", referenced from:
      TensorFlowLite.SignatureRunner.(tensor in _FF0433A885FC97E4F565F3E81A9383E1)(named: Swift.String, withType: TensorFlowLite.TensorType) throws -> TensorFlowLite.Tensor in TensorFlowLite(SignatureRunner.o)
  "_TfLiteSignatureRunnerGetInputTensor", referenced from:
      TensorFlowLite.SignatureRunner.copy(_: Foundation.Data, toInputNamed: Swift.String) throws -> () in TensorFlowLite(SignatureRunner.o)
      TensorFlowLite.SignatureRunner.(tensor in _FF0433A885FC97E4F565F3E81A9383E1)(named: Swift.String, withType: TensorFlowLite.TensorType) throws -> TensorFlowLite.Tensor in TensorFlowLite(SignatureRunner.o)

< MANY MORE LINKER ERRORS FOLLOW >
```

This seems similar to: https://github.com/tensorflow/tensorflow/issues/52042#issuecomment-1098076930

**Comment by sschaetz** [comment]
Update:

I'm able to reproduce this with the [sound classification example](https://github.com/tensorflow/examples/tree/master/lite/examples/sound_classification/ios). Steps to reproduce:

- `git clone https://github.com/tensorflow/examples.git`
- `cd lite/examples/sound_classification/ios`
- `pod update`
- `pod install`
- `open SoundClassification.xcworkspace`
- fix development team
- build -> error "Could not find `TensorFlowLiteSelectTfOps`"
- fix build error by patching project file with 
  ```patch
  --- a/lite/examples/sound_classification/ios/SoundClassification.xcodeproj/project.pbxproj
  +++ b/lite/examples/sound_classification/ios/SoundClassification.xcodeproj/project.pbxproj
  @@ -525,11 +525,11 @@
                                        "-weak_framework",
                                        "\"CoreML\"",
                                        "-force_load",
  -                                       "$(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps",
  +                                       "$(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.xcframework/ios-arm64/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps",
                                );
  @@ -561,11 +561,11 @@
                                        "-weak_framework",
                                        "\"CoreML\"",
                                        "-force_load",
  -                                       "$(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps",
  +                                       "$(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.xcframework/ios-arm64/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps",
- rebuild
- error: [tflite_ios_build_fail.log](https://github.com/tensorflow/tensorflow/files/8770415/tflite_ios_build_fail.log)

**Comment by sschaetz** [comment]
If I force the pods to 0.0.1-nightly.20201122 it works fine. So some time between `20201122` and `20220524` things broke.

**Comment by adamsch1** [comment]
For what it's worth, I hit on this issue.   I also changed the force_load settings as [sschaetz] did above.  Now what I'm faced with are a large set of undefined symbols, mostly around a 3rd party matrix multiplication package for arm called 'ruy'.

Undefined symbol: ruy::ThreadPool::ExecuteImpl(int, int, ruy::Task*)

Additionally it has undefined symbols in tflite:

Undefined symbol: tflite::CheckedLog2(float, int*)

I've tried nightly and  0.0.1-nightly.20201122 for both Pods.   

Any suggestion?

**Comment by adamsch1** [comment]
Note the documentation indicates the wrong path should be used for -force_load

https://www.tensorflow.org/lite/guide/ops_select#ios

-force_load $(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps

**Comment by gdannyliao** [comment]
> For what it's worth, I hit on this issue. I also changed the force_load settings as [sschaetz] did above. Now what I'm faced with are a large set of undefined symbols, mostly around a 3rd party matrix multiplication package for arm called 'ruy'.
> 
> Undefined symbol: ruy::ThreadPool::ExecuteImpl(int, int, ruy::Task*)
> 
> Additionally it has undefined symbols in tflite:
> 
> Undefined symbol: tflite::CheckedLog2(float, int*)
> 
> I've tried nightly and 0.0.1-nightly.20201122 for both Pods.
> 
> Any suggestion?

I just remove the force_load flag and the build is ok

**Comment by dathudeptrai** [comment]
@gdannyliao I also had this issue. My model need to use flexDelegate so i added "-force_load $(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps" to "Other Linker Flags" but when i built an app there was many "Undefined symbols for architecture arm64" bug. Downgrade `TensorFlowLiteSwift ` and `TensorFlowLiteSelectTfOps ` to 2.6.0 work fine on IOS but my model need to use tf 2.9.x to be able to convert GroupConvolution to tflite.

**Comment by dathudeptrai** [comment]
@adamsch1 have you fixed the issue yet?.

**Comment by dathudeptrai** [comment]
@rthadur can you take a look ?

**Comment by sachinprasadhs** [comment]
@sschaetz, Could you try to build with the latest version and check if you're able to build successfully. Refer the document here https://www.tensorflow.org/lite/guide/ops_select#using_bazel_xcode.
If it is failing, as suggested in above comments, could you try building by removing `force_load` flag. Thanks!

**Comment by adamsch1** [comment]
sorry guys I was not able to find a work around.

**Comment by yishuangP** [comment]
Hi really sorry for the late reply. I'm working on a fix, basically we should remove `xnnpack_delegate` target from `avoid_deps`, there are some duplicate symbols in TensorFlowLiteSelectTfOps and TensorFlowLiteC static frameworks, we'll need to remove them from TensorFlowLiteSelectTfOps. In the meantime, can you try tensorflow/lite/ios/build_frameworks.sh? You might need to add `--config=monolithic` when building TensorFlowLiteSelectTfOps_framework.

**Comment by yishuangP** [comment]
Hi, the fix should be in the latest nightly version (0.0.1-nightly.20220830), could you give it a try? Thanks!

**Comment by adamsch1** [comment]
Thanks YishuangP!  Can you share a command/build line for iOS?

**Comment by yishuangP** [comment]
Hi, you just need to update your pod file to depend on the latest nightly build version.

```
...
target 'my-app' do
  use_frameworks!

  # Pods for my-app
  pod 'TensorFlowLiteSwift', '~> 0.0.1-nightly'
  pod 'TensorFlowLiteSelectTfOps', '~> 0.0.1-nightly'
end
```

Or your could hardcode the build version to 0.0.1-nightly.20220830
```
...
target 'my-app' do
  use_frameworks!

  # Pods for my-app
  pod 'TensorFlowLiteSwift', '0.0.1-nightly.20220830'
  pod 'TensorFlowLiteSelectTfOps', '0.0.1-nightly.20220830'
end
```
Then run `pod update` and `pod install`

**Comment by yishuangP** [comment]
Hi sorry we had an issue in our script that builds the xcframework. Submitting the fix, hopefully tonight's build will resolve the issue.

**Comment by yishuangP** [comment]
Hi, I just verified that this build issue is resolved in build `0.0.1-nightly.20220831`. Unfortunately the fix won't be in 2.9.1, if you want to use 2.9.1, you can try building the framework using tensorflow/lite/ios/build_frameworks.sh. Besides, make sure you add this build flag `--define=tflite_with_xnnpack=false` when building the flex delegate framework.

**Comment by johnfhima** [comment]
I still have the same issue.
Is there any fix that exist ?
